ownership concept updated (#92)
exercism/cairo: concepts/ownership/about.md

In Cairo, the concept of ownership is attached to variables rather than the values they hold.
This means that while multiple variables can refer to the same value, ownership is concerned with the variables themselves.
Values in Cairo are immutable, meaning once a value is created, it cannot be changed.
This immutability ensures that even if many variables refer to the same value, the value itself remains constant and unaltered.

 ## The ownership rules.
  _make sure to keep these rules in mind as you work with Cairo_
```
- Each variable in Cairo has an owner.
- There can only be one owner at a time.
- When the owner goes out of scope, the variable is destroyed.
```
### Variable scope
Technically a scope is the range within a program in which a variable is valid

```rust
fn main(){
     let x = 3;
}
   
```
the `x` is an integer which is valid at the point in which it is declared up to the end of the given scope, in this case `main()` function.

```rust
fn main(){ 
                      //x is not valid here; it has not been declared
     let x = 3;       //x is valid from this point forward ; it is been declared
// do stuffs with x   // x is also valid here; it is still in scope

> } // this scope ended here and x goes out of scope and no longer valid.

Reviewer: ```suggestion
} // this scope ended here, `x` goes out of scope and is no longer valid
```

   
```

> Note: 

Reviewer: ```suggestion

Note: 
```
nit: add spacing between separate components, paragraphs, examples etc., makes for a more nicely formatted file

 - When `x` comes into scope, it is valid.
 - It remains valid until it goes out of scope.

>  ### Moving Vaues

Reviewer: ```suggestion
## Moving Values

```
nit: 2nd level subheader, and the space before it is not needed

> A move occurs when a value is passed to another function, rendering the original variable that referred to the value useless. This happens because moving a value causes the original variable to be destroyed, making it unusable.

Reviewer: ```suggestion
A move occurs when a value is passed to another function, destroying the original variable that referred to the value (thus making the variable useless), and creating a new variable to hold the same value.
```

> complex data types are the types of data that is moved when passed into a function and array serves as a good example;

Reviewer: I think it's best we reuse word-for-word the content from [The Cairo Book > Moving values](https://book.cairo-lang.org/ch04-01-what-is-ownership.html#moving-values) - the explanation and the example are very simple, so we should reuse.

Update the rest of this section with The Book's content, starting from:
> Arrays are an example of a complex type that is moved when passing it to another function. So how does the type system ensure that the Cairo program never tries to write to the same memory cell twice? Consider the following code, where we try to remove the front of the array twice:
...

Then insert The Book's example and the paragraph after it. 
You can include the inline comments explaining each line, I like that.

You don't need to add the compilation result section from The Book though, that would be redundant.

let's take a look at it;
```rust
fn get_len(arr: Array<u128>) -> u32{
    arr.len()
}
fn foo(arr: Array<u128>) {
    arr.pop_front();
}

fn main(){
    let mut new_arr: Array<u128> = array![1,2,3,4,5,6,7,8]; // declaration of array
    let len = get_len(new_arr);  //  new_arr value is moved here, rendering the variable useless going forward.
    foo(new_arr); // this will result in bug since new_arr is no longer valid
    let new_len = new_arr.len(); // this also not valid logic, new_arr has been moved out of this scope.
}

```
the above code won't compile because the `new_arr` variable is moved when passed to `get_len()`, rendering `new_arr` unusable afterwards. Subsequent references to `new_arr` result in a compilation error since the original variable is no longer valid after move occur.

> 

Reviewer: Add a short [The Copy Trait](https://book.cairo-lang.org/ch04-01-what-is-ownership.html#the-copy-trait) section

> 

Reviewer: Add a [Destroying Values](https://book.cairo-lang.org/ch04-01-what-is-ownership.html#destroying-values---example-with-feltdict) section

> 

Reviewer: Add a short [Return Values and Scope](https://book.cairo-lang.org/ch04-01-what-is-ownership.html#return-values-and-scope) section

Author: All noted.
I am on it

Author: would you like these other section to contain code sample?

Reviewer: Yes, just keep the examples and descriptions short

Author: HI @misicnenad, I made a commit 2 days ago, not sure you are aware.
would like you to review it.

Reviewer: I noticed you were making commits, but I thought the work was still in progress.
Whenever you're ready for a review, feel free to click on the "re-request review" button at the top of the page (the little circle arrows in the image below):
![image](https://github.com/user-attachments/assets/6a55bec7-7bf0-44a5-a4de-04ffed62f6c1)

Author: oooh Alright, 
yes, I just did that.
Thank you.

> ### Snapshot

Reviewer: Now that I think about it, it would actually make more sense to separate the "Ownership" concept from the "References and Snapshots" concept. This concept is getting too long anyway.

Can you create a new `references-and-snapshots` concept folder and move the rest of this file into its _about.md_?

I created a new issue and linked it to this PR.

Author: sure, I will do that

# Ownership

In Cairo, the concept of ownership is attached to _variables_ rather than the _values_ they hold.

Values in Cairo are immutable, this immutability ensures that even if many variables refer to the same value, the value itself remains constant and unaltered. So the compiler must ensure their correct usage by the programmer. This is the meaning of variable _ownership_: the owner is the code that can read (and write if mutable) the variable.

## Variable Ownership Rules

```
1. Each variable in Cairo has an owner.
2. There can only be one owner at a time.
3. When the owner goes out of scope, the variable is destroyed.
```

## Variable Scope

A scope is the range within a program in which a variable is valid. Let's look at the following example that declares a variable `x` of unsigned integer type, with comments annotating where the variable is valid:

```rust
fn main() {
                      // `x` is not valid here, it has not been declared yet

>     let x: u32 = 3;    // `x` is declared, it is valid from this point forward

Reviewer: ```suggestion
    let x: u32 = 3;   // `x` is declared, it is valid from this point forward
```

                      // do stuff with `x`

} // this scope ended here, `x` goes out of scope and is no longer valid
```

Notes:

- When `x` comes into scope, it is valid.
- It remains valid until it goes out of scope, in this case until the end of the `main()` function.

## Moving Values

A move occurs when a value is passed to another function, destroying the original variable that referred to the value (thus making the variable useless), and creating a new variable to hold the same value.

Arrays are an example of a complex type that is moved when passing it to another function. Here is a short reminder of what an array looks like:

```rust

>  let mut arr: Array<u128> = array![];

Reviewer: ```suggestion
    let mut arr: Array<u128> = array![];
```

## Destroying Values
Linear types can also be utilized by being destroyed, which ensures that the associated resource is correctly released. In Cairo, one type that exhibits this behavior is Felt252Dict. For the sake of provability, these dictionaries must be 'squashed' when they are destructed. This step is crucial but can be easily overlooked. Therefore, to ensure that this important step is not missed, the type system and the compiler enforce the squashing of dictionaries when they are destroyed with the `Destruct` trait.
```rust
#[derive(Destruct)]
struct A {
    dict: Felt252Dict<u128>
}

fn main() {
    A { dict: Default::default() }; // No error here
}
```
when A goes out of scope, its dictionary will be automatically squashed, and the program will compile.
without the `Destruct` trait, the code won't compile. because dictionaries won't be able to squashed.

## Return Values and Scope

> Returning values is equivalent to moving them. the example below shows an example of a function that returns some value.

Reviewer: ```suggestion
Returning values is equivalent to moving them. The example below shows a function that returns some value:
```

    let a3 = takes_and_gives_back(a2);    // a2 is moved into
                                          // takes_and_gives_back, which also
                                          // moves its return value into a3

} // Here, a3 goes out of scope and is dropped. a2 was moved, so nothing
  // happens. a1 goes out of scope and is dropped.

fn gives_ownership() -> A {               // gives_ownership will move its
                                          // return value into the function
                                          // that calls it

    let some_a = A {};                    // some_a comes into scope

    some_a                                // some_a is returned and
                                          // moves ownership to the calling
                                          // function
}

// This function takes an instance some_a of A and returns it
fn takes_and_gives_back(some_a: A) -> A { // some_a comes into scope

    some_a                                // some_a is returned and 
                                          // moves ownership to the calling
                                          // function
}
```

> While this works, moving into and out of every function is a bit tedious. What if we want to let a function use a value but not move the value? It’s quite annoying that anything we pass in also needs to be passed back if we want to use it again, in addition to any data resulting from the body of the function that we might want to return as well.

Reviewer: Let's keep this focused just on _moving_ return values, remove everything after the above example to the end of the file
